File: 2025/day7.rb
#!/usr/bin/ruby
# Sunday, December 07, 2025
# Advent of Code 2025

splits = 0                      # part 1
beams = {}                      # part 2

$stdin.each_line do |line|
  row = line.chomp.split ''
  row.each_with_index do |c, pos|
    if c == 'S'
      beams[pos] = 1
    elsif c == '^'
      if beams[pos]
        splits += 1
        beams[pos+1] = beams.fetch(pos+1, 0) + beams[pos]
        beams[pos-1] = beams.fetch(pos-1, 0) + beams[pos]
        beams.delete pos
      end
    end
  end
end

puts splits
puts beams.values.sum

File: 2025/day7.c
/*
 * Advent of Code 2025, Day 7 (afterwards)
 * Sunday, December 07, 2025
 *
 * cc -Wall -O2 -o day7-after day7.c
 */

#include <stdio.h>

// lines are 142 chars long; the first and last characters are never '^'
#define ROW_SIZE 200

int main (int argc, char **argv) {
  char line[ROW_SIZE];
  int splits = 0;
  unsigned long beams[ROW_SIZE];
  for (int i=0; i<ROW_SIZE; i++)
    beams[i] = 0;

  while (fgets(line, ROW_SIZE, stdin) != NULL) {
    for (int i=0; line[i] != '\0'; i++) {
      if (line[i] == 'S') {
	beams[i] = 1;
      }
      else if (line[i] == '^') {
	if (beams[i]) {
	  splits ++;
	  beams[i-1] += beams[i];
	  beams[i+1] += beams[i];
	  beams[i] = 0;
	}
      }
    }
  }
  printf("part 1: %d\n", splits);
  unsigned long sum = 0;
  for (int i=0; i<ROW_SIZE; i++)
    sum += beams[i];
  printf("part 2: %lu\n", sum);

  return 0;
}
